src/Oscoin/Protocol/Sync/RealWorld.hs

```diff
@@ -67,65 +62,104 @@ newDataFetcher
+                rq0 <- HTTP.parseUrlThrow (formatToString url prefix start end)
+                let rq = rq0 { HTTP.port = fromIntegral (addrPort . nodeHttpApiAddr $ peer) }
+
+                response <- C.http rq httpManager
+                pure (  HTTP.responseBody response
+                     .| mapOutput (\r -> case r of
+                                             API.Ok o  -> o
+                                             API.Err e -> panic e
+                                  ) conduitDecodeCBOR
+                     .| streamList)
+            pure $ join r
 
         -- This implementation is basically 'fetchBlockHeaders' from the spec.
-        -- N.b. Not only does it assumes oscoin#550 is implemented, but we are
-        -- also implementing this not very efficiently, by downloading full
-        -- blocks and throwing away the bodies.
-        SGetBlockHeaders -> \Range{..} -> timed maxAllowedNetworkLatency $ do
-            let url = stext
-                    % "/blocks/by-height?start="
-                    % int @Integer
-                    % "&end="
-                    % int @Integer
-            rq0 <- HTTP.parseUrlThrow (formatToString url prefix start end)
-            let rq = rq0 { HTTP.port = fromIntegral (addrPort . nodeHttpApiAddr $ peer) }
-
-            liftIO $ HTTP.withResponse rq httpManager $ \res -> do
-                cborBlob <- HTTP.brRead (HTTP.responseBody res)
-                let (blks :: OldestFirst [] (Block c tx (Sealed c s))) =
-                        CBOR.deserialise . toS $ cborBlob
-                pure $ map blockHeader blks
+        -- N.b. We are implementing this very inefficiently by downloading full
+        -- blocks and throwing away the bodies, because we don't have first-class
+        -- support to retrieve only the block headers from the HTTP API.
+        SGetBlockHeaders -> \Range{..} -> do
+            r <- timed' GetBlockHeaders peerAddr $ catchNetworkError GetBlockHeaders peerAddr $ do
+                let url = stext
+                        % "/blocks/by-height?start="
+                        % int @Integer
+                        % "&end="
+                        % int @Integer
+                rq0 <- HTTP.parseUrlThrow (formatToString url prefix start end)
+                let rq = rq0 { HTTP.port = fromIntegral (addrPort . nodeHttpApiAddr $ peer) }
+
+                liftIO $ HTTP.withResponse rq httpManager $ \res -> do
+                    cborBlob <- HTTP.brRead (HTTP.responseBody res)
+                    let (blks :: API.Result [Block c tx (Sealed c s)]) =
+                            CBOR.deserialise . toS $ cborBlob
+                    case blks of
+                      API.Ok b -> pure $ map blockHeader b
+                      API.Err err -> throwM $ RequestNetworkError GetTip peerAddr (toS err)
+            pure $ join r
+
+-- | Streams the input list one element at the time.
+streamList :: Monad m => ConduitT [a] a m ()
```

Comment:
Isn't this just `Data.Conduit.Combinators.concat`?

Reply:
Is it! You spotted the `conduit` noob 🙋‍♂

Reply:
98e8dcb6